File: Python/Logger/registry.py
"""
Logger registry for managing singleton logger instances.
"""

from typing import Dict, Optional
from .logger import Logger


class LoggerRegistry:
    """Registry for managing singleton logger instances."""
    
    def __init__(self):
        self._loggers: Dict[str, Logger] = {}
    
    def get_logger(self, name: str) -> Optional[Logger]:
        """Get an existing logger by name."""
        return self._loggers.get(name)
    
    def register_logger(self, name: str, logger: Logger) -> None:
        """Register a logger instance."""
        self._loggers[name] = logger
    
    def close_logger(self, name: str) -> bool:
        """Close all handlers for a logger and remove it from registry."""
        if name in self._loggers:
            logger = self._loggers[name]
            # Close all handlers
            for handler in logger.handlers:
                if hasattr(handler, 'close'):
                    handler.close()
            # Remove from registry
            del self._loggers[name]
            return True
        return False
    
    def close_all(self) -> None:
        """Close all registered loggers."""
        for name in list(self._loggers.keys()):
            self.close_logger(name)
    
    def get_all_loggers(self) -> Dict[str, Logger]:
        """Get all registered loggers."""
        return self._loggers.copy()


# Global registry instance
_registry = LoggerRegistry()


def get_logger(name: str) -> Optional[Logger]:
    """Get a logger instance by name from the registry."""
    return _registry.get_logger(name)


def register_logger(name: str, logger: Logger) -> None:
    """Register a logger instance in the registry."""
    _registry.register_logger(name, logger)


def close_logger(name: str) -> bool:
    """Close a logger and remove it from the registry."""
    return _registry.close_logger(name)


def close_all_loggers() -> None:
    """Close all registered loggers."""
    _registry.close_all()


def get_all_loggers() -> Dict[str, Logger]:
    """Get all registered loggers."""
    return _registry.get_all_loggers() 

File: JavaScript/Logger/registry.js
/**
 * Logger registry for managing singleton logger instances.
 */

import { Logger } from './Logger.js';

/**
 * Registry for managing singleton logger instances.
 */
class LoggerRegistry {
    constructor() {
        this._loggers = new Map();
    }

    /**
     * Get an existing logger by name.
     */
    getLogger(name) {
        return this._loggers.get(name) || null;
    }

    /**
     * Register a logger instance.
     */
    registerLogger(name, logger) {
        this._loggers.set(name, logger);
    }

    /**
     * Close all handlers for a logger and remove it from registry.
     */
    closeLogger(name) {
        if (this._loggers.has(name)) {
            const logger = this._loggers.get(name);
            // Close all handlers
            for (const handler of logger.handlers) {
                if (typeof handler.close === 'function') {
                    handler.close();
                }
            }
            // Remove from registry
            this._loggers.delete(name);
            return true;
        }
        return false;
    }

    /**
     * Close all registered loggers.
     */
    closeAll() {
        for (const name of this._loggers.keys()) {
            this.closeLogger(name);
        }
    }

    /**
     * Get all registered loggers.
     */
    getAllLoggers() {
        return new Map(this._loggers);
    }
}

// Global registry instance
const _registry = new LoggerRegistry();

/**
 * Get a logger instance by name from the registry.
 */
export function getLogger(name) {
    return _registry.getLogger(name);
}

/**
 * Register a logger instance in the registry.
 */
export function registerLogger(name, logger) {
    _registry.registerLogger(name, logger);
}

/**
 * Close a logger and remove it from the registry.
 */
export function closeLogger(name) {
    return _registry.closeLogger(name);
}

/**
 * Close all registered loggers.
 */
export function closeAllLoggers() {
    _registry.closeAll();
}

/**
 * Get all registered loggers.
 */
export function getAllLoggers() {
    return _registry.getAllLoggers();
} 